ocpmodels/models/base.py

```diff
@@ -6,29 +6,209 @@
+        output_targets={},
+        node_embedding_dim=None,
+        edge_embedding_dim=None,
+        _torch_initialized=False,
     ) -> None:
-        super(BaseModel, self).__init__()
-        self.num_atoms = num_atoms
-        self.bond_feat_dim = bond_feat_dim
-        self.num_targets = num_targets
+        if not _torch_initialized:
+            super().__init__()
+
+        self.output_targets = output_targets
+        self.num_targets = len(output_targets)
+
+        self.module_dict = nn.ModuleDict({})
+        for target in output_targets:
+            if self.output_targets[target].get("default_head", False):
+                if "decomposition" in self.output_targets[target]:
+                    # If decomposed, we only need heads for the individual components
+                    continue
+
+                if "irrep_dim" in self.output_targets[target]:
+                    if edge_embedding_dim is None:
+                        raise NotImplementedError(
+                            "Model does not support SO(3) equivariant prediction without edge embeddings."
+                        )
+                    embedding_dim = edge_embedding_dim
+                    output_shape = 1
+                else:
+                    embedding_dim = node_embedding_dim
+                    output_shape = self.output_targets[target].get("shape", 1)
+
+                layers = [
+                    Dense(
+                        embedding_dim,
+                        embedding_dim,
+                        activation="silu",
+                    )
+                    for _ in range(
+                        self.output_targets[target].get("num_layers", 2) - 1
+                    )
+                ]
+
+                layers.append(
+                    Dense(embedding_dim, output_shape, activation=None)
+                )
+
+                self.module_dict[target] = nn.Sequential(*layers)
 
     def forward(self, data):
-        raise NotImplementedError
+        batch = data.batch
+        self.device = data.pos.device
+        self.num_atoms = data.atomic_numbers.shape[0]
+        self.num_systems = data.natoms.shape[0]
+
+        # call declared model forward pass
+        out = self._forward(data)
+        results = {}
+
+        for target in self.output_targets:
+            # Skip parent decomposed targets. The components should exist separately
+            if "decomposition" in self.output_targets[target]:
+                continue
+
+            ### for models that directly return desired property, add directly
+            if target not in self.module_dict:
+                pred = out[target]
+                # squeeze if necessary
+                if len(pred.shape) > 1:
+                    pred = pred.squeeze(dim=1)
+
+                results[target] = pred
+                continue
+
+            assert (
+                target not in out
+            ), f"{target} is explicitly returned, but also has a default head"
```

Comment:
Decided for now to throw an error if the model returns an explicit prediction but we also specified it as default head, since that probably means the user did something wrong. I could be convinced to make this a warning instead (and probably default to the explicitly coded output), if anyone can think of a use case for that scenario.

Reply:
I'm fine with this. We should go through our models to properly wrap energy/force blocks to ensure they arent constructed. This does get tricky for schnet https://github.com/Open-Catalyst-Project/ocp/blob/151be4f5d9ea7d1db8e08443df94349071b13fd5/ocpmodels/models/schnet.py#L115-L117 since we import it from geometric.